src/message/messageActionSheet.js

```diff
@@ -16,18 +20,60 @@ const copyToClipboard = async ({ message, auth }) => {
   Clipboard.setString(rawMessage);
 };
 
-export const actionSheetButtons = [
-  'Narrow to conversation',
-  'Reply',
-  'Copy to clipboard',
-  'Cancel'
+const unMuteTopicHelper = ({ auth, message }) => {
+  unMuteTopic(auth, message.display_recipient, message.subject);
+};
+
+const muteTopicHelper = ({ auth, message }) => {
+  muteTopic(auth, message.display_recipient, message.subject);
+};
+
+const unMuteStreamHelper = ({ auth, message, subscriptions }) => {
+  const sub = subscriptions.find(x => x.name === message.display_recipient);
+  unMuteStream(auth, sub.stream_id);
+};
+
+const muteStreamHelper = ({ auth, message, subscriptions }) => {
+  const sub = subscriptions.find(x => x.name === message.display_recipient);
+  muteStream(auth, sub.stream_id);
+};
+
+const skip = () => false;
+
+const actionSheetButtons = [
+  { title: 'Narrow to conversation', onPress: narrowToConversation },
+  { title: 'Reply', onPress: reply },
+  { title: 'Copy to clipboard', onPress: copyToClipboard },
+  // If skip then covered in constructActionButtons
+  { title: 'Unmute Topic', onPress: unMuteTopicHelper, onlyIf: skip },
+  { title: 'Mute Topic', onPress: muteTopicHelper, onlyIf: skip },
+  { title: 'Mute Stream', onPress: muteStreamHelper, onlyIf: skip },
+  { title: 'Unmute Stream', onPress: unMuteStreamHelper, onlyIf: skip },
+  { title: 'Cancel', onPress: skip, onlyIf: skip },
 ];
 
-export const executeActionSheetAction = ({ buttonIndex, ...props }) => {
-  [
-    narrowToConversation,
-    reply,
-    copyToClipboard,
-    () => {},
-  ][buttonIndex](props);
+export const constructActionButtons = ({ message, auth, narrow, subscriptions, mute }) => {
+  const buttons = actionSheetButtons.filter(x =>
+    !x.onlyIf || x.onlyIf({ message, auth, narrow })).map(x => x.title);
+
+  // These are dependent conditions, hence better if we manage here rather than using onlyIf
+  if (message.type === 'stream') {
+    if (isTopicMuted(message, mute)) {
+      buttons.push('Unmute Topic');
```

Comment:
We will want to implement this a bit differently, so we don't use raw strings (that we need to translate).